flink-streaming-java/src/test/java/org/apache/flink/streaming/api/operators/async/AsyncWaitOperatorTest.java

```diff
@@ -241,6 +246,30 @@ public void timeout(Integer input, ResultFuture<Integer> resultFuture) throws Ex
         }
     }
 
+    /**
+     * Checks if timeout has been called after the element has been completed within the timeout.
+     */
+    private static class TimeoutAfterCompletionTestFunction
+            implements AsyncFunction<Integer, Integer> {
+        static final AtomicBoolean TIMED_OUT = new AtomicBoolean(false);
+
+        @Override
+        public void asyncInvoke(Integer input, ResultFuture<Integer> resultFuture) {
+            ForkJoinPool.commonPool()
+                    .submit(
+                            () -> {
+                                Thread.sleep(TIMEOUT / 2);
```

Comment:
500ms might be low enough to cause test instabilities. Maybe instead of away on a latch that you complete at the end of the unit test?